internal/pipe/ko/ko.go

```diff
@@ -0,0 +1,301 @@
+// Package ko implements the pipe interface with the intent of
+// building OCI compliant images with ko.
+package ko
+
+import (
+	stdcontext "context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/goreleaser/goreleaser/internal/semerrgroup"
+	"golang.org/x/tools/go/packages"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/ko/pkg/build"
+	"github.com/google/ko/pkg/commands/options"
+	"github.com/google/ko/pkg/publish"
+	"github.com/goreleaser/goreleaser/internal/ids"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/goreleaser/pkg/context"
+)
+
+// Pipe that build OCI compliant images with ko.
+type Pipe struct{}
+
+func (Pipe) String() string { return "ko" }
+func (Pipe) Skip(ctx *context.Context) bool {
+	return ctx.SkipKo || len(ctx.Config.Kos) == 0
+}
+
+// Default sets the Pipes defaults.
+func (Pipe) Default(ctx *context.Context) error {
+	ids := ids.New("kos")
+	for _, ko := range ctx.Config.Kos {
+		if err := setConfigDefaults(&ko); err != nil {
+			return err
+		}
+		ids.Inc(ko.ID)
+	}
+	return ids.Validate()
+}
+
+func setConfigDefaults(cfg *config.Ko) error {
+	cfg.Push = true
+
+	if cfg.ID == "" {
+		cfg.ID = "default"
+	}
+
+	if cfg.BaseImage == "" {
+		cfg.BaseImage = "cgr.dev/chainguard/static" // TODO: we can discuss on this
+	}
+
+	return nil
+}
+
+// Run executes the Pipe.
+func (Pipe) Run(ctx *context.Context) error {
+	g := semerrgroup.New(ctx.Parallelism)
+	for _, ko := range ctx.Config.Kos {
+		g.Go(doBuild(ctx, ko))
+	}
+	return g.Wait()
+}
+
+type buildOptions struct {
+	ip                   string
+	main                 string
+	flags                []string
+	env                  []string
+	workingDir           string
+	dockerRepo           string
+	cosignRepo           string
+	platforms            []string
+	baseImage            string
+	tags                 []string
+	sbom                 string
+	ldflags              []string
+	bare                 bool
+	preserverImportPaths bool
+	baseImportPaths      bool
+}
+
+var baseImages sync.Map
+
+func (o *buildOptions) makeBuilder(ctx stdcontext.Context) (*build.Caching, error) {
+	bo := []build.Option{
+		build.WithConfig(map[string]build.Config{
+			o.ip: {
+				Ldflags: o.ldflags,
+				Main:    o.main,
+				Flags:   o.flags,
+				Env:     o.env,
+			},
+		}),
+		build.WithPlatforms(o.platforms...),
+		build.WithBaseImages(func(ctx stdcontext.Context, s string) (name.Reference, build.Result, error) {
+			ref, err := name.ParseReference(o.baseImage)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if cached, found := baseImages.Load(o.baseImage); found {
+				return ref, cached.(build.Result), nil
+			}
+
+			desc, err := remote.Get(ref,
+				remote.WithAuthFromKeychain(authn.DefaultKeychain))
```

Comment:
`ko` uses the smart keychain here, instead of just the default one: https://github.com/ko-build/ko/blob/d869cec943a8c36666d86ff30319577b65402c5e/pkg/commands/config.go